From 9c7d22306b50d250a6d5dab607bb6f592812f676 Mon Sep 17 00:00:00 2001
From: SagarMakhar <sagarmakhar@gmail.com>
Date: Sat, 10 Jul 2021 16:49:30 +0000
Subject: [PATCH] Ignore bpf errors for < 4.9 kernels

Change-Id: I037e4745d5bde12e6edd35203d3d3604ef1ec4c2
Signed-off-by: SagarMakhar <sagarmakhar@gmail.com>
Signed-off-by: rajkale99 <kaleraj.rk@gmail.com>
(cherry picked from commit 8a936bb877b9b2295ca71164db40fe493251197f)
(cherry picked from commit b6e4614686ecd1d9f7e914d46316ca0b22b60cb1)
---
 bpfloader/BpfLoader.cpp | 27 ++++++++++++++++-----------
 1 file changed, 16 insertions(+), 11 deletions(-)

diff --git a/bpfloader/BpfLoader.cpp b/bpfloader/BpfLoader.cpp
index 7a68894..788009e 100644
--- a/bpfloader/BpfLoader.cpp
+++ b/bpfloader/BpfLoader.cpp
@@ -112,17 +112,22 @@ void createSysFsBpfSubDir(const char* const prefix) {
 }
 
 int main() {
-    // Load all ELF objects, create programs and maps, and pin them
-    for (const auto location : locations) {
-        createSysFsBpfSubDir(location.prefix);
-        if (loadAllElfObjects(location.dir, location.prefix) != 0) {
-            ALOGE("=== CRITICAL FAILURE LOADING BPF PROGRAMS FROM %s ===", location.dir);
-            ALOGE("If this triggers reliably, you're probably missing kernel options or patches.");
-            ALOGE("If this triggers randomly, you might be hitting some memory allocation "
-                  "problems or startup script race.");
-            ALOGE("--- DO NOT EXPECT SYSTEM TO BOOT SUCCESSFULLY ---");
-            sleep(20);
-            return 2;
+    bool ebpf_supported = android::base::GetBoolProperty("ro.kernel.ebpf.supported", true);
+
+    if (ebpf_supported) {
+        // Load all ELF objects, create programs and maps, and pin them
+        for (const auto location : locations) {
+            createSysFsBpfSubDir(location.prefix);
+            if (loadAllElfObjects(location.dir, location.prefix) != 0) {
+                ALOGE("=== CRITICAL FAILURE LOADING BPF PROGRAMS FROM %s ===", location.dir);
+                ALOGE("If this triggers reliably, you're probably missing kernel options "
+                      "or patches.");
+                ALOGE("If this triggers randomly, you might be hitting some memory allocation "
+                      "problems or startup script race.");
+                ALOGE("--- DO NOT EXPECT SYSTEM TO BOOT SUCCESSFULLY ---");
+                sleep(20);
+                return 2;
+            }
         }
     }
 
